pageserve/test_serve.py

```diff
@@ -18,9 +18,21 @@
 
 
 class TestServe(unittest.TestCase):
-    def test_get_auth(self):
-        # TODO: test authorization retrieval
-        pass
+    def test_auth_json(self):
```

Comment:
So right now this test is more of an integration test between pageserve and users, which sometimes causes it to fail in Travis, any suggestions for restructuring the get_auth_json function in a better way?

Reply:
Given the current structure, mocks with injection to replace the requests module are probably the best way to test this.

Reply:
The other option would be to use something like https://github.com/pretenders/pretenders, which looks to let you set up your own HTTP server, but this part may need to be integration tested.

Reply:
Still a little confused what exactly I'd be testing with a mock - is it just that the get_auth_json goes through all the necessary steps? or would I still test the returned value?

Reply:
You could mock `requests` for the scope of this function and make `.get` return to give you what you want. Using `unittest.mock`, something like
```python
mock_requests = mock.MagicMock()
mock_requests.get().return_value = '{"edit_access":"true"}'
with mock.patch('app.requests', mock_requests):
    response = app.get_auth_json('example_user')
valid_response = response['edit_access'] or not response['edit_access']
self.assertTrue(valid_response)
```

Reply:
You can write the context manager as:

```python
with mock.patch('app.requests') as mock_requests:
  mock_requests.get().return_value = '{"edit_access":"true"}'
  response = app.get_auth_json('example_user')
...
```

Reply:
@evankanderson what's the best way to handle the r.json() function here? it's expecting to be called on a response from the requests library and I'm not sure how to patch both functions at once

Reply:
```suggestion
with mock.patch('app.requests') as mock_requests:
    mock_response = mock.MagicMock()
    mock_response.json().return_value = '{"edit_access":"true"}'
    mock_requests.get().return_value = mock_response
    response = app.get_auth_json('example_user')
```

Reply:
That did it, thanks!

Reply:
I did have to take change json() to json - same for post().